src/lib/tls/tls13_pqc/hybrid_public_key.cpp

```diff
@@ -0,0 +1,373 @@
+      return AlgorithmIdentifier(spec.second, AlgorithmIdentifier::USE_EMPTY_PARAM);
+   });
+
+   return result;
+}
+
+std::vector<size_t> public_value_lengths_for_group(Group_Params group) {
+   BOTAN_ASSERT_NOMSG(is_hybrid(group));
+
+   // This duplicates information of the algorithm internals.
+   //
+   // TODO: Find a way to expose important algorithm constants globally
+   //       in the library, to avoid violating the DRY principle.
+   switch(group) {
+      case Group_Params::HYBRID_X25519_KYBER_512_R3_CLOUDFLARE:
+      case Group_Params::HYBRID_X25519_KYBER_512_R3_OQS:
+         return {32, 800};
+      case Group_Params::HYBRID_X25519_KYBER_768_R3_CLOUDFLARE:
+      case Group_Params::HYBRID_X25519_KYBER_768_R3_OQS:
+         return {32, 1184};
+
+      case Group_Params::HYBRID_SECP256R1_KYBER_512_R3_OQS:
+         return {32, 800};
+      case Group_Params::HYBRID_SECP384R1_KYBER_768_R3_OQS:
+         return {48, 1184};
+      case Group_Params::HYBRID_SECP521R1_KYBER_1024_R3_OQS:
+         return {66, 1568};
+
+      default:
+         return {};
+   }
+}
+
+}  // namespace
+
+std::unique_ptr<Hybrid_KEM_PublicKey> Hybrid_KEM_PublicKey::load_for_group(
+   Group_Params group, std::span<const uint8_t> concatenated_public_values) {
+   auto public_value_lengths = public_value_lengths_for_group(group);
+   auto alg_ids = algorithm_identifiers_for_group(group);
+   BOTAN_ASSERT_NOMSG(public_value_lengths.size() == alg_ids.size());
+
+   BufferSlicer public_value_slicer(concatenated_public_values);
+   std::vector<std::unique_ptr<Public_Key>> pks;
+   for(size_t idx = 0; idx < alg_ids.size(); ++idx) {
+      pks.emplace_back(load_public_key(alg_ids[idx], public_value_slicer.take(public_value_lengths[idx])));
+   }
+   BOTAN_ASSERT(public_value_slicer.empty(), "Concatenated public value has the correct length");
+   return std::make_unique<Hybrid_KEM_PublicKey>(std::move(pks));
+}
+
+Hybrid_KEM_PublicKey::Hybrid_KEM_PublicKey(std::vector<std::unique_ptr<Public_Key>> pks) {
+   BOTAN_ARG_CHECK(!pks.empty(), "List of public keys must not be empty");
+   BOTAN_ARG_CHECK(std::all_of(pks.begin(), pks.end(), [](const auto& pk) { return pk != nullptr; }),
+                   "List of public keys contains a nullptr");
+   BOTAN_ARG_CHECK(std::all_of(pks.begin(),
+                               pks.end(),
+                               [](const auto& pk) {
+                                  return pk->supports_operation(PublicKeyOperation::KeyEncapsulation) ||
+                                         pk->supports_operation(PublicKeyOperation::KeyAgreement);
+                               }),
+                   "Some provided public key is not compatible with this hybrid wrapper");
+
+   std::transform(
+      pks.begin(), pks.end(), std::back_inserter(m_public_keys), [](auto& key) -> std::unique_ptr<Public_Key> {
+         if(key->supports_operation(PublicKeyOperation::KeyAgreement)) {
```

Comment:
If a key supports both key agreement and key encapsulation, we will use `KEX_to_KEM_Adapter_PublicKey` - is that really the right choice?